src/transformations/defmodel.cpp

```diff
@@ -0,0 +1,445 @@
+                return false;
+            }
+            checkedHorizontal = true;
+        }
+        return true;
+    }
+
+    bool getLonLatOffset(int ix, int iy, double &lonOffsetRadian,
+                         double &latOffsetRadian) const {
+        if (!checkHorizontal(STR_DEGREE)) {
+            return false;
+        }
+        float lonOffsetDeg;
+        float latOffsetDeg;
+        if (!realGrid->valueAt(ix, iy, sampleX, lonOffsetDeg) ||
+            !realGrid->valueAt(ix, iy, sampleY, latOffsetDeg)) {
+            return false;
+        }
+        lonOffsetRadian = lonOffsetDeg * DEG_TO_RAD;
+        latOffsetRadian = latOffsetDeg * DEG_TO_RAD;
+        return true;
+    }
+
+    bool getZOffset(int ix, int iy, double &zOffset) const {
+        if (!checkedVertical) {
+            const auto samplesPerPixel = realGrid->samplesPerPixel();
+            if (samplesPerPixel == 1) {
+                sampleZ = 0;
+            } else if (samplesPerPixel < 3) {
+                pj_log(ctx, PJ_LOG_ERROR,
+                       "defmodel: grid %s has not enough samples",
+                       realGrid->name().c_str());
+                return false;
+            }
+            bool foundDesc = false;
+            bool foundDescZ = false;
+            for (int i = 0; i < samplesPerPixel; i++) {
+                const auto desc = realGrid->description(i);
+                if (desc == "vertical_offset") {
+                    sampleZ = i;
+                    foundDescZ = true;
+                }
+                if (!desc.empty()) {
+                    foundDesc = true;
+                }
+            }
+            if (foundDesc && !foundDescZ) {
+                pj_log(ctx, PJ_LOG_ERROR,
+                       "defmodel: grid %s : Found band description, "
+                       "but not the ones expected",
+                       realGrid->name().c_str());
+                return false;
+            }
+            const auto unit = realGrid->unit(sampleZ);
+            if (!unit.empty() && unit != STR_METRE) {
+                pj_log(ctx, PJ_LOG_ERROR,
+                       "defmodel: grid %s : Only unit=metre currently "
+                       "handled for this mode",
+                       realGrid->name().c_str());
+                return false;
+            }
+            checkedVertical = true;
+        }
+        float zOffsetFloat = 0.0f;
+        const bool ret = realGrid->valueAt(ix, iy, sampleZ, zOffsetFloat);
+        zOffset = zOffsetFloat;
+        return ret;
+    }
+
+    bool getEastingNorthingOffset(int ix, int iy, double &eastingOffset,
+                                  double &northingOffset) const {
+        if (!checkHorizontal(STR_METRE)) {
+            return false;
+        }
+        float eastingOffsetFloat = 0.0f;
+        float northingOffsetFloat = 0.0f;
+        const bool ret =
+            realGrid->valueAt(ix, iy, sampleX, eastingOffsetFloat) &&
```

Comment:
Are the grids always floats, or could they be doubles too?

Reply:
We can support other data types in the TIFF storage (see GTiffGrid::valueAt()) but we normalize to float. Given that the values are corrections, and not absolute values, float is accurate enough. This is as a heritage of formats like GTX or NTv2 storing floats.